content/posts/aws-autoscaling-groups-with-terraform.md

```diff
@@ -0,0 +1,193 @@
+
+> Throughout the examples we omitted some configuration elements creation for brevity.
+> Others are injected as variables and are also not shown here.
+
+## Initial Configuration
+First create the launch template
+```terraform
+# Query for the latest Ubuntu 18.04 AMI in the region
+data "aws_ami" "ubuntu" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*"]
+  }
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
+
+  owners = ["099720109477"] # Canonical
+}
+
+resource "aws_launch_template" "mysvc" {
+  name                                 = "mysvc"
+  ebs_optimized                        = true
+  image_id                             = data.aws_ami.ubuntu.id
+  instance_initiated_shutdown_behavior = "terminate"
+  instance_type                        = "c5.large"
+  key_name                             = var.key_name
+  network_interfaces {
+    associate_public_ip_address = true
+    subnet_id                   = var.subnet_id
+    security_groups             = var.security_groups
+    delete_on_termination       = true
+  }
+  tag_specifications {
+    resource_type = "instance"
+    tags = {
+      Name          = "mysvc"
+    }
+  }
+  user_data = base64encode(templatefile("${path.module}/user_data.yaml", {
+    web_password    = var.web_password,
+    ssl_private_key = tls_private_key.mysvc.private_key_pem,
+    ssl_certificate = tls_self_signed_cert.mysvc.cert_pem
+  }))
+
+  # we don't want to create a new template just because there is a newer AMI
+  lifecycle {
+    ignore_changes = [
+      image_id,
+    ]
+  }
+}
+
+```
+The launch template defines the critical elements of the instance but also loads
+the configuration of the service from the user data file.  The user data file is a template
+that evaluates to a valid cloud-init configuration file.
+
+Now let's create the autoscaling group
+```terraform
+resource "aws_autoscaling_group" "mysvc" {
+  name                      = "mysvc-${aws_launch_template.mysvc.latest_version}"
+  health_check_type         = "ELB"
+  health_check_grace_period = 120
+  termination_policies      = ["OldestInstance"]
+  launch_template {
+    id      = aws_launch_template.mysvc.id
+    version = aws_launch_template.mysvc.latest_version
+  }
+  min_size = 1
+  max_size = 10
+
+  lifecycle {
+    create_before_destroy = true
+  }
+  target_group_arns = [aws_lb_target_group.svc.arn]
+}
+```
+There are a few things to note here:
+1. We name the autoscaling group with the latest version of the launch template.  This is to make
+sure that when the launch template is updated we are creating a new autoscaling group.
+2. When creating a new autoscaling group we want to first create the new one and then destroy
+the previous one.  Hence the life-cycle directive `create_before_destroy = true`. As our
+service is behind an ELB this will make sure no down time during the configuration change.
```

Comment:
Indent this so it's rendered correctly as a list.

Reply:
Ouch, I needed the preview for that.  Note that in the GitHub preview this looks as a list.
Also the other colouring issue you mentioned in Slack.